### ISSUE TITLE
Install fails unless --locked option is set

### ISSUE THREAD

**Issue by JamesHallowell** [cargo-public-api/cargo-public-api#61]
Hi!

This looks like a really great tool. I just tried to install it using the instructions in the README and hit the following issue:

```
$ cargo install cargo-public-api
    Updating crates.io index
  Downloaded cargo-public-api v0.12.0
  Downloaded 1 crate (14.4 KB) in 0.28s
  Installing cargo-public-api v0.12.0
  Downloaded clap_derive v3.2.6
  Downloaded clap v3.2.6
  Downloaded unicode-ident v1.0.1
  Downloaded indexmap v1.9.1
  Downloaded proc-macro2 v1.0.40
  Downloaded clap_lex v0.2.3
  Downloaded hashbrown v0.12.1
  Downloaded cargo_toml v0.11.5
  Downloaded anyhow v1.0.58
  Downloaded diff v0.1.12
  Downloaded public-api v0.12.0
  Downloaded hashbag v0.1.6
  Downloaded syn v1.0.98
  Downloaded quote v1.0.20
  Downloaded rustdoc-types v0.11.0
  Downloaded 15 crates (876.4 KB) in 0.93s
   Compiling proc-macro2 v1.0.40
   Compiling unicode-ident v1.0.1
   Compiling quote v1.0.20
   Compiling syn v1.0.98
   Compiling serde_derive v1.0.137
   Compiling serde v1.0.137
   Compiling version_check v0.9.4
   Compiling autocfg v1.1.0
   Compiling libc v0.2.126
   Compiling serde_json v1.0.81
   Compiling itoa v1.0.2
   Compiling ryu v1.0.10
   Compiling semver v1.0.10
   Compiling camino v1.0.9
   Compiling hashbrown v0.12.1
   Compiling heck v0.4.0
   Compiling os_str_bytes v6.1.0
   Compiling anyhow v1.0.58
   Compiling strsim v0.10.0
   Compiling bitflags v1.3.2
   Compiling termcolor v1.1.3
   Compiling textwrap v0.15.0
   Compiling hashbag v0.1.6
   Compiling once_cell v1.12.0
   Compiling diff v0.1.12
   Compiling ansi_term v0.12.1
   Compiling proc-macro-error-attr v1.0.4
   Compiling proc-macro-error v1.0.4
   Compiling indexmap v1.9.1
   Compiling clap_lex v0.2.3
   Compiling atty v0.2.14
   Compiling thiserror-impl v1.0.31
   Compiling clap_derive v3.2.6
   Compiling thiserror v1.0.31
   Compiling clap v3.2.6
   Compiling toml v0.5.9
   Compiling rustdoc-types v0.11.0
   Compiling cargo-platform v0.1.2
   Compiling cargo_metadata v0.14.2
   Compiling cargo_toml v0.11.5
   Compiling public-api v0.12.0
error[E0271]: type mismatch resolving `<impl Iterator<Item = (&PublicItem, usize)> as Iterator>::Item == &PublicItem`
   --> /Users/james.hallowell/.cargo/registry/src/github.com-1ecc6299db9ec823/public-api-0.12.0/src/diff.rs:70:48
    |
70  |         let mut removed_paths: ItemsWithPath = bag_to_path_map(all_removed);
    |                                                ^^^^^^^^^^^^^^^ expected `&PublicItem`, found tuple
    |
    = note: expected reference `&PublicItem`
                   found tuple `(&PublicItem, usize)`
note: required by a bound in `bag_to_path_map`
   --> /Users/james.hallowell/.cargo/registry/src/github.com-1ecc6299db9ec823/public-api-0.12.0/src/diff.rs:116:47
    |
116 | fn bag_to_path_map<'a>(hashbag: impl Iterator<Item = &'a PublicItem>) -> ItemsWithPath {
    |                                               ^^^^^^^^^^^^^^^^^^^^^ required by this bound in `bag_to_path_map`

error[E0271]: type mismatch resolving `<impl Iterator<Item = (&PublicItem, usize)> as Iterator>::Item == &PublicItem`
   --> /Users/james.hallowell/.cargo/registry/src/github.com-1ecc6299db9ec823/public-api-0.12.0/src/diff.rs:71:46
    |
71  |         let mut added_paths: ItemsWithPath = bag_to_path_map(all_added);
    |                                              ^^^^^^^^^^^^^^^ expected `&PublicItem`, found tuple
    |
    = note: expected reference `&PublicItem`
                   found tuple `(&PublicItem, usize)`
note: required by a bound in `bag_to_path_map`
   --> /Users/james.hallowell/.cargo/registry/src/github.com-1ecc6299db9ec823/public-api-0.12.0/src/diff.rs:116:47
    |
116 | fn bag_to_path_map<'a>(hashbag: impl Iterator<Item = &'a PublicItem>) -> ItemsWithPath {
    |                                               ^^^^^^^^^^^^^^^^^^^^^ required by this bound in `bag_to_path_map`

For more information about this error, try `rustc --explain E0271`.
error: could not compile `public-api` due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
error: failed to compile `cargo-public-api v0.12.0`, intermediate artifacts can be found at `/var/folders/2j/z687lxf542q2syccnrr_2fx80000gn/T/cargo-installrUofUg
```

I had a look in the CI config for this repo and noticed the `--locked` flag is being passed in, so tried setting that, which did let me install the tool successfully:

```
$ cargo install cargo-public-api --locked
    Updating crates.io index
  Installing cargo-public-api v0.12.0
  Downloaded clap_derive v3.2.5
  Downloaded hashbag v0.1.5
  Downloaded clap_lex v0.2.2
  Downloaded indexmap v1.9.0
  Downloaded clap v3.2.5
  Downloaded 5 crates (313.7 KB) in 0.39s
   Compiling proc-macro2 v1.0.40
   Compiling unicode-ident v1.0.1
   Compiling quote v1.0.20
   Compiling syn v1.0.98
   Compiling serde_derive v1.0.137
   Compiling serde v1.0.137
   Compiling version_check v0.9.4
   Compiling autocfg v1.1.0
   Compiling libc v0.2.126
   Compiling serde_json v1.0.81
   Compiling itoa v1.0.2
   Compiling camino v1.0.9
   Compiling semver v1.0.10
   Compiling ryu v1.0.10
   Compiling hashbrown v0.12.1
   Compiling anyhow v1.0.58
   Compiling heck v0.4.0
   Compiling os_str_bytes v6.1.0
   Compiling hashbag v0.1.5
   Compiling textwrap v0.15.0
   Compiling termcolor v1.1.3
   Compiling bitflags v1.3.2
   Compiling once_cell v1.12.0
   Compiling strsim v0.10.0
   Compiling diff v0.1.12
   Compiling ansi_term v0.12.1
   Compiling proc-macro-error-attr v1.0.4
   Compiling proc-macro-error v1.0.4
   Compiling indexmap v1.9.0
   Compiling clap_lex v0.2.2
   Compiling atty v0.2.14
   Compiling thiserror-impl v1.0.31
   Compiling clap_derive v3.2.5
   Compiling thiserror v1.0.31
   Compiling clap v3.2.5
   Compiling rustdoc-types v0.11.0
   Compiling toml v0.5.9
   Compiling cargo-platform v0.1.2
   Compiling cargo_metadata v0.14.2
   Compiling cargo_toml v0.11.5
   Compiling public-api v0.12.0
   Compiling cargo-public-api v0.12.0
    Finished release [optimized] target(s) in 27.74s
  Installing /Users/james.hallowell/.cargo/bin/cargo-public-api
   Installed package `cargo-public-api v0.12.0` (executable `cargo-public-api`)
```

One difference appears to be the version of hashbag, it succeeds with hashbag v0.1.5, but failed on v0.1.6. Having a dig around in that repo it looks like possibly its something related to this change - https://github.com/jonhoo/hashbag/commit/4b9fe778dbc58fffd38a093f0b9fec0a781706c9 - based on the compiler error I hit.

Let me know if I can provide any further information that would be useful.

Thanks!

**Comment by Enselic** [comment]
Thank you so much for reporting this 🙏 

This is caused by my `DifferenceAddon::difference()` having the same name as `HashBag::difference()`. If I change `DifferenceAddon::difference()`  to `DifferenceAddon::difference2()`, the code compiles. I am surprised Rust resolves the method to the intrinsic method when `DifferenceAddon` is still in scope. But it is probably a known quirk, and I have only myself to blame.

Note to self: Next time you vendor code, make sure to use a unique method name to avoid future conflicts with the upstream release! Or use explicit syntax such as:
```rust
        let all_removed = DifferenceAddon::difference(&old, &new);
        let all_added = DifferenceAddon::difference(&new, &old);
```

I have now released v0.12.1 so that `--locked` is not needed any longer. Please let me know if you still have problems!

**Comment by JamesHallowell** [comment]
Thanks @Enselic! Just tried to install it again and v0.12.1 worked perfectly.
